From 97c2cbb181a45c3443fb434ef7ca2928b780f337 Mon Sep 17 00:00:00 2001
From: Pritesh Patel <pritesh.patel@einfochips.com>
Date: Fri, 19 Jul 2024 06:45:16 +0000
Subject: [PATCH 010/128] drivers: eswin: Enable ways for ccache

Enabled all available cache ways to be used as cache.
This change is required for CPU performance improvement.

Out of reset, only way 0 is enabled and the disabled ways
are addressable in L2-LIM(Loosely Integrated Memory).

Signed-off-by: Pritesh Patel <pritesh.patel@einfochips.com>
---
 drivers/soc/sifive/sifive_ccache.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/drivers/soc/sifive/sifive_ccache.c b/drivers/soc/sifive/sifive_ccache.c
index a0bf6d3747e6..928ab5dcd3c3 100644
--- a/drivers/soc/sifive/sifive_ccache.c
+++ b/drivers/soc/sifive/sifive_ccache.c
@@ -253,6 +253,9 @@ static int __init sifive_ccache_init(void)
 	int i, rc, intr_num;
 	const struct of_device_id *match;
 	unsigned long quirks;
+#ifdef CONFIG_SOC_SIFIVE_EIC7700
+	unsigned int config, ways;
+#endif
 
 	np = of_find_matching_node_and_match(NULL, sifive_ccache_ids, &match);
 	if (!np)
@@ -294,6 +297,12 @@ static int __init sifive_ccache_init(void)
 	}
 	of_node_put(np);
 
+#ifdef CONFIG_SOC_SIFIVE_EIC7700
+	config = readl(ccache_base + SIFIVE_CCACHE_CONFIG);
+	ways = (config >> 8) & 0xff;
+	writel(ways-1, ccache_base + SIFIVE_CCACHE_WAYENABLE);
+#endif
+
 #ifdef CONFIG_RISCV_NONSTANDARD_CACHE_OPS
 	if (quirks & QUIRK_NONSTANDARD_CACHE_OPS) {
 			riscv_cbom_block_size = SIFIVE_CCACHE_LINE_SIZE;
-- 
2.47.0

